text/0000-lifecycle-pkg-refactor

```diff
@@ -0,0 +1,174 @@
+* CNB business logic should live inside the `buildpack`, `lifecycle`, and `platform` packages (and their nested packages) or their `internal` equivalents. All other packages that are direct children of `pkg` or `internal` should be CNB-free.
+* Logic that branches on api version should be handled by an internal service nested under the appropriate package - e.g., logic that branches on buildpack api should live in `internal/buildpack` - see [example](https://github.com/buildpacks/lifecycle/blob/b6803be364429689a818a5d0a09db42bd21b9995/buildpack/bom.go#L14-L23).
+* Tests should serve as documentation for apis by enumerating which services are in use - see [example](https://github.com/buildpacks/lifecycle/blob/a457244e2785afd342fe511563445a4fa4554e3e/platform/inputs/analyzer_factory_test.go#L353-L375).
+* To keep the launcher as lean as possible, we should minimize the packages it depends on (hence `platform/launch/inputs`).
+
+## Aligning with the spec
+
+The spec has this structure (abridged):
+
+Platform:
+
+  - Stacks
+  - Lifecycle
+    - Operations
+      - Build
+      - Rebase
+      - Launch
+    - Platform inputs and outputs (args, flags, env vars, mounts, exit codes)
+    - Registry auth
+  - Buildpacks directory layout
+  - Additional guidance
+    - Environment
+      - Stack provided vars
+      - Platform provided vars (platform/env)
+    - Caching
+  - Data format (file and label schema)
+
+Buildpack:
+
+  - Buildpack interface
+    - Detect
+    - Build
+    - Exec.d
+    - Layer types
+  - App interface (profile.d)
+  - Detection: /bin/detect, order resolution
+  - Analysis: restored metadata
+  - Build: /bin/build, build plan, SBOM, layers
+  - Export: layer re-use, slices, caching
+  - Launch: execution strategy (direct/indirect), exec.d,  profile.d
+  - Environment
+    - Layer env dirs
+    - Platform provided (stack, platform/env)
+    - Modification rules (append, prepend, etc.)
+  - Data format (file and exec.d schema)
+
+# Migration
+[migration]: #migration
+
+Lifecycle contributors will need to agree on and become comfortable using the new structure. Library consumers will need to update package import paths in their code.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Changing package import paths will be a breaking change for library consumers (GitHub [search](https://github.com/search?o=desc&q=%22github.com%2Fbuildpacks%2Flifecycle%22&s=indexed&type=Code)). However this is mitigated by version locking in `go.mod`.
+
+There have also (to this point) been no firm stability guarantees when using the lifecycle as a library. If we (hopefully, through this RFC) settle on a package structure that works for us, we might be able to offer some promises of stability in the future.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- What other designs have been considered? Doing nothing, just keep adding code as best we can.
+- Why is this proposal the best? Making it easier to add new code will allow us to deliver features faster.
```

Comment:
Another benefit of having a well-organized lifecycle is that by adding good README files and docs comments, we can turn the lifecycle into an easy-to-consume repository of information for further understanding the CNB spec.